src/tls/cockpit-certificate-ensure.c

```diff
@@ -72,6 +72,33 @@ read_file (const char     *filename,
   close (fd);
 }
 
+static void
+write_file (int                   dirfd,
+            const char           *dirfd_filename,
+            const char           *filename,
+            const gnutls_datum_t *data,
+            uid_t                 uid,
+            gid_t                 gid)
+{
+  /* Just open the file directly: it doesn't exist yet and nobody will
+   * look at it until after we're done here.
+   */
+  int fd = openat (dirfd, filename, O_CREAT | O_EXCL | O_WRONLY, 0400);
+
+  if (fd == -1)
+    err (EXIT_FAILURE, "%s/%s: creat", dirfd_filename, filename);
+
+  if (fchown (fd, uid, gid) != 0)
+    err (EXIT_FAILURE, "%s/%s: fchown", dirfd_filename, filename);
+
+  size_t s = write (fd, data->data, data->size);
+  if (s == -1)
+    err (EXIT_FAILURE, "%s/%s: write", dirfd_filename, filename);
+  if (s != data->size)
+    errx (EXIT_FAILURE, "%s/%s: write: wrote %zu bytes, expecting %zu",
+          dirfd_filename, filename, s, (size_t) data->size);
+}
```

Comment:
Missing `close (fd)`

Reply:
Disagree.  I'm not worrying about closing/freeing in the cases where I call `err()`, otherwise this would all get quite a lot larger.

Reply:
I did not mean closing the file in the failure case, but at the end of the function in the success case. Sure, c-c-ensure is not a long-lived process, so the fd leak is not a big issue -- but I'd still appreciate if write_file() does not leak the fd.

Reply:
Ah!  Absolutely.  I misunderstood the patch context.